cannot convert ‘[...] ndarray<bool, 1>’ to ‘bool’ in assignment

The following code fails to compile:

```# pythran export weights_pythran2(uint8[:, :], float)
def weights_pythran2(input_data, threshold=0.3):
    n_seq, length = input_data.shape
    weights = np.zeros(n_seq, dtype=np.float32)

    for i in range(n_seq):
        vector = input_data[i, None, :]
        count_matches = np.sum(vector == input_data, axis=1)
        over_threshold = count_matches > (threshold * length)
        total = np.sum(over_threshold)
        weights[i] = 1 / total

    return weights
```

Pythran spits a lot, but the last message is: `/usr/include/c++/7/bits/stl_algo.h:4306:12: error: cannot convert ‘{anonymous}::pythonic::types::ndarray<bool, 1>’ to ‘bool’ in assignment`

I cannot find exactly where the error happens, since if I remove the assignment to weights[i], Pythran will optimise out everything from there on. Do you have any tip to locate the offending line?

I am using Python 3.6 on Linux with GCC 7.2. An alternative version of the code works:

```# pythran export weights_pythran(int8[:, :], float)
def weights_pythran(input_data, threshold=0.3):
    n_seq, length = input_data.shape
    counts = np.sum(input_data == input_data[:, None, :], axis=2)
    distances = np.sum(counts > (threshold * length), axis=1)
    return 1 / distances
```

--- a/pythran/pythonic/include/types/numpy_gexpr.hpp
+++ b/pythran/pythonic/include/types/numpy_gexpr.hpp
@@ -57,12 +57,14 @@ namespace types
   template <size_t C>
   struct extended_slice {
     template <class T, size_t N, class... S>
-    numpy_gexpr<ndarray<T, N + C>, typename to_slice<S>::type...>
-    operator()(ndarray<T, N> &&a, S const &... s);
+    auto operator()(ndarray<T, N> &&a, S const &... s)
+        -> decltype(std::declval<ndarray<T, N + C>>()(
+            std::declval<typename to_slice<S>::type>()...));
 
     template <class T, size_t N, class... S>
-    numpy_gexpr<ndarray<T, N + C>, typename to_slice<S>::type...>
-    operator()(ndarray<T, N> const &a, S const &... s);
+    auto operator()(ndarray<T, N> const &a, S const &... s)
+        -> decltype(std::declval<ndarray<T, N + C>>()(
+            std::declval<typename to_slice<S>::type>()...));
   };
 
   template <>

--- a/pythran/pythonic/types/numpy_gexpr.hpp
+++ b/pythran/pythonic/types/numpy_gexpr.hpp
@@ -131,24 +131,24 @@ namespace types
    */
   template <size_t C>
   template <class T, size_t N, class... S>
-  numpy_gexpr<ndarray<T, N + C>, typename to_slice<S>::type...>
-      extended_slice<C>::operator()(ndarray<T, N> &&a, S const &... s)
+  auto extended_slice<C>::operator()(ndarray<T, N> &&a, S const &... s)
+      -> decltype(std::declval<ndarray<T, N + C>>()(
+          std::declval<typename to_slice<S>::type>()...))
   {
-    return {
-        std::move(a).reshape(make_reshape<N + C>(
-            a.shape(), array<bool, sizeof...(S)>{to_slice<S>::is_new_axis...})),
-        to_slice<S>{}(s)...};
+    return std::move(a).reshape(make_reshape<N + C>(
+        a.shape(), array<bool, sizeof...(S)>{to_slice<S>::is_new_axis...}))(
+        to_slice<S>{}(s)...);
   }
 
   template <size_t C>
   template <class T, size_t N, class... S>
-  numpy_gexpr<ndarray<T, N + C>, typename to_slice<S>::type...>
-      extended_slice<C>::operator()(ndarray<T, N> const &a, S const &... s)
+  auto extended_slice<C>::operator()(ndarray<T, N> const &a, S const &... s)
+      -> decltype(std::declval<ndarray<T, N + C>>()(
+          std::declval<typename to_slice<S>::type>()...))
   {
-    return {a.reshape(make_reshape<N + C>(
-                a.shape(),
-                array<bool, sizeof...(S)>{{to_slice<S>::is_new_axis...}})),
-            to_slice<S>{}(s)...};
+    return a.reshape(make_reshape<N + C>(
+        a.shape(), array<bool, sizeof...(S)>{{to_slice<S>::is_new_axis...}}))(
+        to_slice<S>{}(s)...);
   }
 
   template <class T, size_t N, class... S>

--- a/pythran/tests/cases/weights.py
+++ b/pythran/tests/cases/weights.py
@@ -1,0 +1,16 @@
+#pythran export weights(uint8[:,:])
+#runas import numpy as np; x = np.arange(10, dtype=np.uint8).reshape(5,2); weights(x)
+
+import numpy as np
+def weights(input_data, threshold=0.3):
+    n_seq, length = input_data.shape
+    weights = np.zeros(n_seq, dtype=np.float32)
+
+    for i in range(n_seq):
+        vector = input_data[i, None, :]
+        count_matches = np.sum(vector == input_data, axis=1)
+        over_threshold = count_matches > (threshold * length)
+        total = np.sum(over_threshold)
+        weights[i] = 1 / total
+
+    return weights

